pkg/notify/notifier/pushover/pushover.go

```diff
@@ -0,0 +1,225 @@
+package pushover
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/kubesphere/notification-manager/pkg/async"
+	"github.com/kubesphere/notification-manager/pkg/notify/config"
+	"github.com/kubesphere/notification-manager/pkg/notify/notifier"
+	"github.com/kubesphere/notification-manager/pkg/utils"
+	"github.com/prometheus/alertmanager/template"
+)
+
+const (
+	DefaultSendTimeout = time.Second * 3
+	URL                = "https://api.pushover.net/1/messages.json"
+	DefaultTemplate    = `{{ template "nm.default.text" . }}`
+	PoMsgLimitAlert    = 99
+)
+
+type Notifier struct {
+	notifierCfg  *config.Config
+	pushover     []*config.Pushover
+	timeout      time.Duration
+	logger       log.Logger
+	template     *notifier.Template
+	templateName string
+	client       http.Client
+}
+
+type pushoverRequest struct {
+	pushoverMessage
+}
+
+type pushoverResponse struct {
+	Status  int      `json:"status"`
+	Request string   `json:"request"`
+	Errors  []string `json:"errors,omitempty"`
+	Receipt string   `json:"receipt,omitempty"`
+}
+
+func NewPushoverNotifier(logger log.Logger, receivers []config.Receiver, notifierCfg *config.Config) notifier.Notifier {
+
+	var path []string
+	opts := notifierCfg.ReceiverOpts
+	if opts != nil && opts.Global != nil {
+		path = opts.Global.TemplateFiles
+	}
+	tmpl, err := notifier.NewTemplate(path)
+	if err != nil {
+		_ = level.Error(logger).Log("msg", "PushoverNotifier: get template error", "error", err.Error())
+		return nil
+	}
+
+	n := &Notifier{
+		notifierCfg:  notifierCfg,
+		timeout:      DefaultSendTimeout,
+		logger:       logger,
+		template:     tmpl,
+		templateName: DefaultTemplate,
+		client:       http.Client{Timeout: DefaultSendTimeout},
+	}
+
+	if opts != nil && opts.Pushover != nil {
+
+		if opts.Pushover.NotificationTimeout != nil {
+			n.timeout = time.Second * time.Duration(*opts.Pushover.NotificationTimeout)
+			n.client.Timeout = n.timeout
+		}
+
+		if len(opts.Pushover.Template) > 0 {
+			n.templateName = opts.Pushover.Template
+		} else if opts.Global != nil && len(opts.Global.Template) > 0 {
```

Comment:
If the `opts.Pushover`, can this code be executed? Please refer to this code.
https://github.com/kubesphere/notification-manager/blob/master/pkg/notify/notifier/slack/slack.go#L65

Reply:
Resolved: https://github.com/kubesphere/notification-manager/blob/6229e77a79cb5cadc513d62513f96706f9879ebc/pkg/notify/notifier/pushover/pushover.go#L74